rfcs/20190117-tf-module.md

```diff
@@ -0,0 +1,634 @@
+
+### Non-functional
+
+**Wide adoption**
+
+-   `tf.Module` (or types derived from it) should be the obvious choice for
+    research code and useful in general.
+-   It should not favor a specific theme (e.g. supervised learning) vs. another
+    (e.g. adversarial training).
+-   Having shared abstractions is critical for a research teams to facilitate
+    easy collaboration.
+
+**Researcher approved**
+
+-   `tf.Module` should be designed in collaboration with researchers and the API
+    should match their preferences.
+
+**Unbloated**
+
+-   `tf.Module` should be a thin abstraction which is trivial to reason about.
+-   The implementation should be clean and readable, ideally no more than a few
+    100 lines of code.
+-   The external API should have very few public methods.
+
+**Tightly specified**
+
+-   The specification for a `tf.Module` should be clear and agreed upon.
+-   Future additions to `tf.Module` should be critically evaluated against the
+    original design goals and moved elsewhere when they do not align.
+-   New features (e.g. supporting mixed precision training) should not require
+    changes to `tf.Module`.
+
+**Loosely coupled**
+
+-   Using `tf.Module` should not tie you to a particular framework or mentality.
+-   It should be trivial to integrate a subclass of `tf.Module` into any
+    codebase.
+
+**Defensive**
+
+-   The design of `tf.Module` should optimize for helping avoid common mistakes
+    (e.g. accidentally not tracking variables).
+-   If functionality is required (e.g. calling `super(..).__init__()` in
+    subclasses) we should error early (e.g. immediately after the module
+    constructor returns) to remind users if they forget.
+
+## Detailed Design
+
+### Class structure
+
+We propose a pure Python class `tf.Module` which has a metaclass
+`tf.ModuleMetaclass` to wrap all methods in the modules name scope. This design
+allows for submodules or variables to created in any method (including
+`__init__` and `__call__`). There is more detail on this decision in the
+detailed design below.
+
+```python
+class Module(six.with_metaclass(ModuleMetaclass, tf.checkpoint.Checkpointable)):
+  """Base neural network module class.
+
+  A module is a named container for `tf.Variable`s, other `tf.Module`s and
+  functions which apply to user input. For example a dense layer in a neural
+  network might be implemented as a `tf.Module`:
+
+  >>> class Dense(tf.Module):
+  ...   def __init__(self, in_features, output_features):
+  ...     super(Linear, self).__init__()
+  ...     # NOTE: We're creating variables in the ctor, but you can actually
+  ...     #       create them in any method (including __call__) you like.
+  ...     self.w = tf.Variable(
+  ...         tf.random.normal([input_features, output_features]), name='w')
+  ...     self.b = tf.Variable(tf.zeros([output_features]), name='b')
+  ...
+  ...   def __call__(self, x):
+  ...     x = tf.convert_to_tensor(x, name='x')
```

Comment:
I feel it is important to avoid having the users to call `tf.convert_to_tensor` every time they define their own `__call__` function. One might want to define implicit pre-call hook to do the conversion automatically. Or better approach might be to explicitly require the API to only accept `Tensor` (or nested tuple of tensors) as inputs, and ask user to do explicit conversion before calling from top level.

This is because many of the modules are probably to be used as building blocks in bigger architectures (e.g. residual blocks inside a resnet). They will not directly interact with the user most of the time, so there is probably no need for them to call `convert_to_tensor` every time, but it might be better to enforce the API to have consistent behavior.

Reply:
While I agree with you that it would be good for any callable subclass of `tf.Module` to have some automatic conversion to tensor (e.g., in an implementation of Layer 2.0), this is beyond the scope of `tf.Module`. The `Dense` class example only serves to demonstrate how one would subclass `tf.Module`. Let me try to lay out what I understand the purpose of `tf.Module` to be and what the reason for its limited scope is. (I am not affiliated with the TF team)

After the shift to object-oriented storing of `tf.Variable`'s, variable reuse is done by storing variables within object attributes. It would be impractical and clunky to store all variables in a unique container object (which would essentially recreate TF 1.0's collections). To avoid this, we prefer to have variables contained within the objects that would use them, e.g., `kernel` and `bias` variables as attributes of the `Dense` instance that uses it.  While this new way of organizing variables has many advantages, it does make checkpointing and operating on the "global" variables more difficult (e.g., collecting variables for gradient descent).

By design, `tf.Module` is lightweight and focuses on addressing the challenges of managing nested stateful objects. Its sole purpose is to abstract away the DAG of modules for the purpose of operating on all relevant variables. This allows us to act like our top-level/root module explicitly contains all our variables. This is done by providing traversal methods which collect all the variables. As a bonus, `tf.Module` automatically handles the complexities of checkpointing when variables are allowed to be created before and after the checkpoint is loaded, e.g., hydration and instantiation of variables.

By limiting `tf.Module`'s responsibilities to managing nested stateful modules, we end up with code that is easy to understand, document and extend. As a result, any developer wishing to implement some stateful containers can easily do so without intricate knowledge of how best to manage "global" state by subclassing `tf.Module`. Since handling stateful DAGs of modules is a very common requirement, `tf.Module` could easily be used by most sub-packages and libraries with stateful operations, e.g., `tf.optimizers`, `tf.metrics`, `tf.estimator`. (Note, this specific proposal is arguing for the existence/inclusion of `tf.Module` in TF 2.0. The advantages of using `tf.Module` as a base class throughout tensorflow will likely be examined on case by case basis in a separate proposal.)

Reply:
@gehring Yes, I agree with you. I don't think doing automatic conversion is a good idea. But enforcing the inputs are Tensors are probably good (even only implicit as a statement in the documentation). Anyhow, this is a minor thing in the whole proposal.

Reply:
I'm strongly against auto-conversion to tensors and agree with @tomhennigan, it is up to a scientists, engineers or anyone else, what to do with non tensor types.